File: programming-basics/Python/bachelor/year-1/semester-2/lab-4/TEquation.py
class TEquation:
    _number_of_coefficients = 0
    _coefficients = []
    _roots = []

    def __init__(self, *coefficients):
        if self.are_coefficients_valid(coefficients):
            self._coefficients = coefficients
        else:
            raise ValueError("Invalid coefficients: {}".format(coefficients))

    def are_coefficients_valid(self, coefficients):
        is_valid_quantity = len(self.get_coefficients()) == self.get_number_of_coefficients()
        return is_valid_quantity and coefficients[0] != 0

    def find_roots(self):
        pass

    def get_roots(self):
        return self._roots

    def get_coefficients(self):
        return self._coefficients

    def get_number_of_coefficients(self):
        return self._number_of_coefficients

    def is_root(self, root):
        if self.get_roots() is None:
            return False

        if len(self.get_roots()) == 0:
            self.find_roots()

        for r in self.get_roots():
            if r == root:
                return True

    def calculate_roots_sum(self):
        if self.get_roots() is None:
            return None

        if len(self.get_roots()) == 0:
            self.find_roots()

        return sum(self.get_roots())

File: programming-basics/C-plus-plus/bachelor/year-1/semester-2/lab-4/TEquation.cpp
#include "TEquation.h"

TEquation::TEquation(): coefficients(std::vector<double>()) {}

TEquation::TEquation(std::vector<double> coeff, int coeffsNum) {
    if (this->areCoefficientsValid(coeff, coeffsNum)) {
        this->coefficients = coeff;
    } else {
        throw "Invalid coefficients";
    }
}

bool TEquation::areCoefficientsValid(std::vector<double> coeffs) {
    bool isValidLength = coeffs.size() == this->getNumberOfCoefficients();
    return isValidLength && coefficients[0] != 0;
}

bool TEquation::areCoefficientsValid(std::vector<double> coeffs, int coeffsNum) {
    bool isFirstValid = coeffs[0] != 0;
    bool isValidLength = coeffs.size() == coeffsNum;
    return isValidLength && isFirstValid;
}

std::vector<double> TEquation::findRoots() {
    return std::vector<double>();
}

std::vector<double> TEquation::getRoots() {
    return this->roots;
}

std::vector<double> TEquation::getCoefficients() {
    return this->coefficients;
}

int TEquation::getNumberOfCoefficients() {
    return this->numberOfCoefficients;
}

bool TEquation::isRoot(double root) {
    if (this->roots.size() == 0) {
        return false;
    }

    for (auto r : this->roots) {
        if (r == root) {
            return true;
        }
    }
    return false;
}

double TEquation::calculateRootsSum() {
    double sum = 0;
    for (auto r : this->roots) {
        sum += r;
    }
    return sum;
}

void TEquation::print() {
    std::cout << "Coefficients: [";
    for (auto c: this->coefficients) {
        std::cout << c << " ";
    }
    std::cout << "]" << std::endl;

    std::cout << "Roots: [";
    for (auto r: this->roots) {
        std::cout << r << " ";
    }
    std::cout << "]" << std::endl;
};
